bson/src/main/org/bson/codecs/AbstractMapCodec.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2017 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs;
+
+import org.bson.BsonBinarySubType;
+import org.bson.BsonReader;
+import org.bson.BsonType;
+import org.bson.BsonWriter;
+import org.bson.Transformer;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import static java.util.Arrays.asList;
+import static org.bson.assertions.Assertions.notNull;
+import static org.bson.codecs.configuration.CodecRegistries.fromProviders;
+
+abstract class AbstractMapCodec<T extends Map<String, Object>>  implements Codec<T> {
+    private static final CodecRegistry DEFAULT_REGISTRY = fromProviders(asList(new ValueCodecProvider(), new BsonValueCodecProvider(),
+            new DocumentCodecProvider(), new IterableCodecProvider(), new MapCodecProvider()));
+    private static final BsonTypeClassMap DEFAULT_BSON_TYPE_CLASS_MAP = new BsonTypeClassMap();
+    private final BsonTypeCodecMap bsonTypeCodecMap;
+    private final CodecRegistry registry;
+    private final Transformer valueTransformer;
+
+    AbstractMapCodec() {
+        this(DEFAULT_REGISTRY);
+    }
+
+    AbstractMapCodec(final CodecRegistry registry) {
+        this(registry, DEFAULT_BSON_TYPE_CLASS_MAP);
+    }
+
+    AbstractMapCodec(final CodecRegistry registry, final BsonTypeClassMap bsonTypeClassMap) {
+        this(registry, DEFAULT_BSON_TYPE_CLASS_MAP, null);
+    }
+
+    AbstractMapCodec(final CodecRegistry registry, final BsonTypeClassMap bsonTypeClassMap, final Transformer valueTransformer) {
+        this.registry = notNull("registry", registry);
+        this.bsonTypeCodecMap = new BsonTypeCodecMap(notNull("bsonTypeClassMap", bsonTypeClassMap), registry);
+        this.valueTransformer = valueTransformer != null ? valueTransformer : new Transformer() {
+            @Override
+            public Object transform(final Object value) {
+                return value;
+            }
+        };
+    }
+
+    abstract T newInstance();
+
+    abstract void beforeFields(BsonWriter bsonWriter, EncoderContext encoderContext, T map);
+
+    abstract boolean skipField(EncoderContext encoderContext, String key);
+
+    void encodeMap(final BsonWriter writer, final T map, final EncoderContext encoderContext) {
+        writeMap(writer, map, encoderContext);
+    }
+
+    T decodeMap(final BsonReader reader, final DecoderContext decoderContext) {
```

Comment:
You can exclude it in clirr-exclude.yml:

    org.bson.codecs.DocumentCodec:
    - decode(org.bson.BsonReader,org.bson.codecs.DecoderContext)